### ISSUE TITLE
Crypto.Symmetric.Poly1305.poly1305_mac

### ISSUE THREAD

**Issue by nikswamy** [FStarLang/FStar#792]
Crypto.Symmetric.Poly1305.fst is in our nightly regression suite. 

It has often been failing, although it did succeed on Dec 20th.
https://raw.githubusercontent.com/project-everest/ci-logs/master/fstar-20161220020039.all

The problematic proof is this one, which has been failing consistently for me.
```
(.\Crypto.Symmetric.Poly1305.fst(1099,94-1133,50))
        Query (Crypto.Symmetric.Poly1305.poly1305_mac, 2)       failed in 642752 milliseconds with fuel 1 and ifuel 1
```

However, in that successful Dec 20th build, this particular query went through. Quoting from the log:

```
(./Crypto.Symmetric.Poly1305.fst(1097,59-1131,50))
Query (Crypto.Symmetric.Poly1305.poly1305_mac, 2)	succeeded in 618075 milliseconds with fuel 1 and ifuel 1
```

Note, the line numbers are slightly different since I have some local options setting the fuel to 1, 1 to match the successful run. 

It seems like this particular proof needs some rewriting for stability. Your help is much appreciated. 

Doing a run now with --detail_errors to see if I can get something more meaningful

**Comment by nikswamy** [comment]
here's the detailed error report:

```
Detailed error report follows for .\Crypto.Symmetric.Poly1305.fst(1099,94-1133,50)
Taking 5 seconds per proof obligation (95 proofs in total)
OK: proof obligation at .\..\..\..\ulib\prims.fst(642,17-642,23) was proven
OK: proof obligation at .\Crypto.Symmetric.Poly1305.Bigint.fst(26,27-26,50) was proven
OK: proof obligation at .\Crypto.Symmetric.Poly1305.Bigint.fst(202,34-202,42) was proven
OK: proof obligation at .\Crypto.Symmetric.Poly1305.Bigint.fst(202,55-202,70) was proven
OK: proof obligation at .\Crypto.Symmetric.Poly1305.Bigint.fst(203,68-203,86) was proven
OK: proof obligation at .\Crypto.Symmetric.Poly1305.Spec.fst(38,23-38,40) was proven
OK: proof obligation at .\Crypto.Symmetric.Poly1305.Spec.fst(38,23-38,40) was proven
OK: proof obligation at .\Crypto.Symmetric.Poly1305.fst(49,33-49,56) was proven
OK: proof obligation at .\Crypto.Symmetric.Poly1305.fst(49,33-49,56) was proven
OK: proof obligation at .\Crypto.Symmetric.Poly1305.fst(49,33-49,56) was proven
OK: proof obligation at .\Crypto.Symmetric.Poly1305.fst(49,33-49,56) was proven
OK: proof obligation at .\Crypto.Symmetric.Poly1305.fst(49,33-49,56) was proven
OK: proof obligation at .\Crypto.Symmetric.Poly1305.fst(49,33-49,56) was proven
OK: proof obligation at .\Crypto.Symmetric.Poly1305.fst(49,33-49,56) was proven
OK: proof obligation at .\Crypto.Symmetric.Poly1305.fst(52,21-52,35) was proven
OK: proof obligation at .\Crypto.Symmetric.Poly1305.fst(52,21-52,35) was proven
OK: proof obligation at .\Crypto.Symmetric.Poly1305.fst(53,24-53,38) was proven
OK: proof obligation at .\Crypto.Symmetric.Poly1305.fst(53,24-53,38) was proven
OK: proof obligation at .\Crypto.Symmetric.Poly1305.fst(53,24-53,38) was proven
OK: proof obligation at .\Crypto.Symmetric.Poly1305.fst(60,31-60,39) was proven
OK: proof obligation at .\Crypto.Symmetric.Poly1305.fst(60,31-60,39) was proven
OK: proof obligation at .\Crypto.Symmetric.Poly1305.fst(60,31-60,39) was proven
OK: proof obligation at .\Crypto.Symmetric.Poly1305.fst(60,31-60,39) was proven
OK: proof obligation at .\Crypto.Symmetric.Poly1305.fst(60,31-60,39) was proven
OK: proof obligation at .\Crypto.Symmetric.Poly1305.fst(60,31-60,39) was proven
OK: proof obligation at .\Crypto.Symmetric.Poly1305.fst(93,31-93,39) was proven
OK: proof obligation at .\Crypto.Symmetric.Poly1305.fst(93,31-93,39) was proven
OK: proof obligation at .\Crypto.Symmetric.Poly1305.fst(93,31-93,39) was proven
OK: proof obligation at .\Crypto.Symmetric.Poly1305.fst(630,16-630,28) was proven
OK: proof obligation at .\Crypto.Symmetric.Poly1305.fst(631,15-631,31) was proven
OK: proof obligation at .\Crypto.Symmetric.Poly1305.fst(631,35-631,49) was proven
OK: proof obligation at .\Crypto.Symmetric.Poly1305.fst(631,53-631,67) was proven
OK: proof obligation at .\Crypto.Symmetric.Poly1305.fst(633,22-633,30) was proven
OK: proof obligation at .\Crypto.Symmetric.Poly1305.fst(633,34-633,42) was proven
OK: proof obligation at .\Crypto.Symmetric.Poly1305.fst(633,46-633,56) was proven
OK: proof obligation at .\Crypto.Symmetric.Poly1305.fst(633,60-633,83) was proven
OK: proof obligation at .\Crypto.Symmetric.Poly1305.fst(712,22-712,32) was proven
OK: proof obligation at .\Crypto.Symmetric.Poly1305.fst(712,36-712,46) was proven
OK: proof obligation at .\Crypto.Symmetric.Poly1305.fst(712,50-712,58) was proven
OK: proof obligation at .\Crypto.Symmetric.Poly1305.fst(713,7-713,17) was proven
OK: proof obligation at .\Crypto.Symmetric.Poly1305.fst(714,7-714,23) was proven
OK: proof obligation at .\Crypto.Symmetric.Poly1305.fst(714,27-714,41) was proven
OK: proof obligation at .\Crypto.Symmetric.Poly1305.fst(714,45-714,59) was proven
OK: proof obligation at .\Crypto.Symmetric.Poly1305.fst(1018,13-1018,32) was proven
OK: proof obligation at .\Crypto.Symmetric.Poly1305.fst(1019,15-1019,31) was proven
OK: proof obligation at .\Crypto.Symmetric.Poly1305.fst(1020,13-1020,27) was proven
OK: proof obligation at .\Crypto.Symmetric.Poly1305.fst(1020,31-1020,45) was proven
OK: proof obligation at .\Crypto.Symmetric.Poly1305.fst(1022,24-1022,34) was proven
OK: proof obligation at .\Crypto.Symmetric.Poly1305.fst(1022,38-1022,48) was proven
OK: proof obligation at .\Crypto.Symmetric.Poly1305.fst(1022,52-1022,60) was proven
OK: proof obligation at .\Crypto.Symmetric.Poly1305.fst(1022,64-1022,83) was proven
OK: proof obligation at .\Crypto.Symmetric.Poly1305.fst(1065,15-1065,32) was proven
OK: proof obligation at .\Crypto.Symmetric.Poly1305.fst(1066,16-1066,33) was proven
OK: proof obligation at .\Crypto.Symmetric.Poly1305.fst(1067,16-1067,33) was proven
OK: proof obligation at .\Crypto.Symmetric.Poly1305.fst(1068,16-1068,26) was proven
OK: proof obligation at .\Crypto.Symmetric.Poly1305.fst(1069,16-1069,33) was proven
OK: proof obligation at .\Crypto.Symmetric.Poly1305.fst(1070,16-1070,32) was proven
OK: proof obligation at .\Crypto.Symmetric.Poly1305.fst(1071,16-1071,26) was proven
OK: proof obligation at .\Crypto.Symmetric.Poly1305.fst(1071,30-1071,40) was proven
OK: proof obligation at .\Crypto.Symmetric.Poly1305.fst(1071,44-1071,54) was proven
OK: proof obligation at .\Crypto.Symmetric.Poly1305.fst(1072,16-1072,38) was proven
OK: proof obligation at .\Crypto.Symmetric.Poly1305.fst(1073,16-1073,35) was proven
OK: proof obligation at .\Crypto.Symmetric.Poly1305.fst(1074,16-1074,38) was proven
OK: proof obligation at .\Crypto.Symmetric.Poly1305.fst(1075,16-1075,28) was proven
OK: proof obligation at .\Crypto.Symmetric.Poly1305.fst(1090,24-1090,34) was proven
OK: proof obligation at .\Crypto.Symmetric.Poly1305.fst(1090,38-1090,48) was proven
OK: proof obligation at .\Crypto.Symmetric.Poly1305.fst(1090,52-1090,62) was proven
OK: proof obligation at .\Crypto.Symmetric.Poly1305.fst(1091,29-1091,40) was proven
OK: proof obligation at .\Crypto.Symmetric.Poly1305.fst(1091,44-1091,55) was proven
OK: proof obligation at .\Crypto.Symmetric.Poly1305.fst(1091,59-1091,70) was proven
OK: proof obligation at .\Crypto.Symmetric.Poly1305.fst(1092,9-1092,29) was proven
OK: proof obligation at .\Crypto.Symmetric.Poly1305.fst(1096,9-1098,56) was proven
.\Crypto.Symmetric.Poly1305.fst(1116,2-1116,8): (Error) assertion failed (see .\Crypto.Symmetric.Poly1305.fst(1116,9-1116,31))
OK: proof obligation at .\Crypto.Symmetric.Poly1305.fst(1125,9-1125,70) was proven
OK: proof obligation at .\Crypto.Symmetric.Poly1305.fst(1127,4-1128,75) was proven
OK: proof obligation at .\Crypto.Symmetric.Poly1305.fst(1132,9-1132,56) was proven
OK: proof obligation at .\FStar.Buffer.fst(57,28-57,36) was proven
OK: proof obligation at .\FStar.Buffer.fst(838,45-838,67) was proven
OK: proof obligation at .\FStar.Buffer.fst(838,45-838,67) was proven
OK: proof obligation at .\FStar.Buffer.fst(839,18-839,41) was proven
OK: proof obligation at .\FStar.Buffer.fst(839,18-839,41) was proven
OK: proof obligation at D:\workspace\FStar\ulib\FStar.UInt.fst(24,28-24,36) was proven
OK: proof obligation at D:\workspace\FStar\ulib\FStar.UInt.fst(24,28-24,36) was proven
OK: proof obligation at D:\workspace\FStar\ulib\FStar.UInt.fst(24,28-24,36) was proven
OK: proof obligation at D:\workspace\FStar\ulib\FStar.UInt.fst(24,28-24,36) was proven
OK: proof obligation at D:\workspace\FStar\ulib\FStar.UInt.fst(24,28-24,36) was proven
OK: proof obligation at D:\workspace\FStar\ulib\FStar.UInt.fst(24,28-24,36) was proven
OK: proof obligation at D:\workspace\FStar\ulib\FStar.UInt.fst(24,28-24,36) was proven
OK: proof obligation at D:\workspace\FStar\ulib\FStar.UInt.fst(24,28-24,36) was proven
OK: proof obligation at D:\workspace\FStar\ulib\hyperstack\FStar.ST.fst(30,91-30,109) was proven
OK: proof obligation at D:\workspace\FStar\ulib\hyperstack\FStar.ST.fst(71,52-71,73) was proven
OK: proof obligation at D:\workspace\FStar\ulib\hyperstack\FStar.ST.fst(71,52-71,73) was proven
OK: proof obligation at D:\workspace\FStar\ulib\hyperstack\FStar.ST.fst(112,22-112,32) was proven
OK: proof obligation at unknown(0,0-0,0) was proven
OK: proof obligation at unknown(0,0-0,0) was proven
Verified module: Crypto.Symmetric.Poly1305 (2132356 milliseconds)
1 error was reported (see above)
```

**Comment by jkzinzindohoue** [comment]
This goes through on my machine when setting both init and max fuel to 0, with rlimit at 100. Can someone else reproduce ?

**Comment by jkzinzindohoue** [comment]
Also was there a particular reason why the initial fuel was set to 1 and the max as high as 10 ? In my experience those values have to be kept as low as possible, and both equal if possible. There is no induction anywhere here.

**Comment by aseemr** [comment]
@jkzinzindohoue, it works for me too (in the c_record_planA branch) with the options you mention.

**Comment by jkzinzindohoue** [comment]
Thanks. I suspect that changing the fuel settings for the functions above would also cut down the verification time quite a bit. I will wait for a nightly green and then close the issue.

**Comment by aseemr** [comment]
@jkzinzindohoue, can you please do these changes in the master branch (or merge your changes from c_record_planA into master)? We were thinking of deleting the c_record_planA branch, it's merged into master already.

**Comment by s-zanella** [comment]
Fixed in 28d19dfec20d61a1024792ff90fe6dc547f866f0
